### ISSUE TITLE
Improve display on HiDPI screens

### ISSUE THREAD

**Issue by MattBlissett** [activityworkshop/GpsPrune#16]
With Java 9, and on Linux setting `export GDK_SCALE=2` it's already the case that the interface icons and text are scaled correctly for a HiDPI screen. (Before Java 9, they can be unusably small.) It would be good if the environment variable didn't need to be set, hopefully this can be detected automatically.

Also, the map itself doesn't take advantage of the better screen — lines are chunky, and tiles the default standard resolution ones. Maps look much nicer with HiDPI tiles, there are examples here: https://wiki.openstreetmap.org/wiki/High-resolution_tiles

![no-scaling](https://user-images.githubusercontent.com/583320/34916500-8d019038-f939-11e7-8240-90e7e3b508af.png)

![scaling](https://user-images.githubusercontent.com/583320/34916501-923a405e-f939-11e7-9fb7-bfcf136dd9ad.png)

The two images show with and without the GDK_SCALE variable set.

**Comment by activityworkshop** [comment]
If an environment variable needs to be set, then I think the launch script (maybe /usr/bin/gpsprune if you have one of those?) would be the right place to set it.  That way, everything is set already when java launches the window.
But I'm afraid I have no idea how the launch script would know whether to set the variable or not - would it need to somehow recognise the physical screen size in order to calculate the dpi?

About the map tiles, if I understand correctly then GpsPrune shouldn't have to _handle_ the tiles any differently (they're still 256x256 pixels, just with different rendering parameters to make everything thicker and bigger and more readable).  So I guess you just added a different map source to GpsPrune just like any other map source?  Or did you generate those tiles yourself?  From your link I couldn't find an obvious "high-resolution" worldwide mapnik source, for example.  Again, I don't think it's really GpsPrune's job to decide for you whether you want "high-resolution" versions of tiles or not, it should just let you choose any map source you want.

Another issue I guess is GpsPrune's painting _onto_ the map, like its line-drawing and waypoint name-drawing (and waypoint icons with v19).  Do these need to be modified too, or are they already covered by your "GDK_SCALE=2" environment variable?

**Comment by MattBlissett** [comment]
The environment variable is an easy workaround on Linux.  On Mac OS with a high-resolution screen, the latest GpsPrune interface displays correctly without any changes. I mention it so anyone else with the same issue finds the fix.

* The tiles are more interesting.

A tile for a normal display is 256×256px, and assumes each pixel is around 0.28mm, so the tile covers 71mm of screen. The map designer plans line widths, font sizes and the density of information around this.

Pixels on a double density screen are around 0.14mm, so a tile covering 71×71mm would need to be 512x512px.

This is a standard tile:
<img src="https://tiles.lyrk.org/ls/14/8803/5375?apikey=e9f8eb3824344d18a5b4b657773caf30">

This is a hi-res tile, 512px wide but squashed with `width=256` in HTML — meaning the width of 256 standard pixels:
<img src="https://tiles.lyrk.org/lr/14/8803/5375?apikey=e9f8eb3824344d18a5b4b657773caf30" width="256" height="256">

Both are 71mm wide, but on my hi-res screen, the standard tile has been scaled up to cover 512×512 small pixels -- it's blurry. The second tile is crisp.

The same comparison is made here: https://mc.bbbike.org/mc/?num=2&mt0=lyrk-retina&mt1=lyrk — if you don't have access to a hi-res screen, the issue will become a lot clearer if you look at this page next time you're passing an Apple store or similar.

Note that simply showing four tiles from the next zoom level in isn't appropriate — the text and lines would then be too small to see comfortably. It would look like this:

<img src="https://tiles.lyrk.org/ls/15/17606/10750?apikey=e9f8eb3824344d18a5b4b657773caf30" width="128" height="128"><img src="https://tiles.lyrk.org/ls/15/17607/10750?apikey=e9f8eb3824344d18a5b4b657773caf30" width="128" height="128">
<img src="https://tiles.lyrk.org/ls/15/17606/10751?apikey=e9f8eb3824344d18a5b4b657773caf30" width="128" height="128"><img src="https://tiles.lyrk.org/ls/15/17607/10751?apikey=e9f8eb3824344d18a5b4b657773caf30" width="128" height="128">

With `GDK_SCALE=2` (and on Mac OS), GpsPrune is showing 256px tiles scaled to cover 512 small pixels, so the map is blurry. The lines and markers are also scaled in the same way, and are blurry.

With the default `GDK_SCALE=1`, GpsPrune is showing 256px covering 256 small pixels, so the map is crisp but unreadable. The lines are sharp, but presumably half the intended real width.

I have no idea how it's done in a Java app, but it would need to use the 512px tiles to cover the smaller pixels without scaling.

This almost certainly isn't a quick fix, and isn't practical to work on without a hi-res screen — if you get one, you will want to fix it immediately to take advantage of the screen. GpsPrune works, and it's still common for graphics in some applications or on websites to appear slightly blurry where support is lacking, especially on Windows.  HiDPI screens are becoming more common though, all but the cheapest Apple computers for a few years have them, and in the last year or two many high-end PCs.

**Comment by MattBlissett** [comment]
> it's much more like a spectrum of screen sizes

Yes, indeed. I couldn't find public statistics, though there are some examples here: https://mydevice.io/devices/

On work's tile server I see about 5% of requests are for 2x tiles, less than 0.5% for 3x, less than 0.05% for 4x. Our website maps request `floor(screenPixelDensity)` resolution, so I'm missing the fairly common 1.5x density.

> I understand that zooming the standard 256x256 tiles by a factor of two will make them blurry. But I'm not convinced that downloading 512x512 tiles, then painting them to a 256x256 section of the graphics buffer, and then painting that to the screen with the same automatic zoom factor of two would give any more acceptable results.

It doesn't -- it's worse, with the scaling happening twice. As you write later, the automatic zoom factor would need to be overridden for the map canvas, to enable painting 512x512px tiles to a 512x512 area of the double resolution screen.

> Rather than inventing a new tile format I would suggest that what you really want to do is to paint the 256x256 tiles without the two-times scaling so that they remain sharp. And then select a suitable source of your map tiles which renders the maps (at standard 256x256 size) using larger font sizes, thicker lines, larger symbols and maybe fewer features.

I tried JOSM with the advice on your link, and the default tiles.  The map labels are too small, but they are sharp. (This is about 11cm wide on my screen.)

![josm](https://user-images.githubusercontent.com/583320/35249712-4a47bc22-ffd3-11e7-9de0-1486a74fe605.png)

The cost is 4x as many HTTP requests, which I think is the reason no-one is providing tiles this way. "My" format seems to be more common: [Thunderforest.com](http://www.thunderforest.com) (an OSM developer) has commercial tiles for 2x screens, [Lyrk](https://mc.bbbike.org/mc/?num=2&mt0=lyrk-retina&mt1=lyrk) do too, so do Mapbox.

Mapbox have gone a step further in reducing HTTP requests by changing their standard tiles to 512px, to reduce the number of HTTP requests: https://blog.mapbox.com/512-map-tiles-cb5bfd6e72ba -- they have a 2x version, which is therefore 1024px wide.

I've avoided showing my own work, to show that I'm not inventing anything here. But:

> Such a source could use Mapnik's "scale-factor", which is designed to do exactly this

I've done that, though using 512px tile as the base size like Mapbox.

Tile 6/31/21 is southwest Britain in the Mapbox scheme. Here it is, rendered with increasing size, and increasing scale factor:

512 and scale-factor 1: https://tile.gbif.org/3857/omt/6/31/21@1x.png?style=osm-bright

1024 and scale-factor 2: https://tile.gbif.org/3857/omt/6/31/21@2x.png?style=osm-bright

2048 and scale-factor 3: https://tile.gbif.org/3857/omt/6/31/21@3x.png?style=osm-bright

The advantage of the @1x, @2x increasing resolution and size is that the tile address is constant. The scale factor is [used here](https://github.com/gbif/gbif-basemaps/blob/master/raster-basemap-server/src/main/node/renderer.js#L37). The result is very easy for web developers to use, since the web browser handles the complexity of the "pixel density": [view source this](http://tile.gbif.org/ui/3857/minimal.html).

> Also the Arch wiki page makes everything sound far more complicated than I ever realised it was!

Amen to that.

> if the GDK_SCALE is set to 2 (to make the window decorations and side-bars readable), then is it possible to switch this off again for the main map panel so that the scaling doesn't make it all blurry? If I understand you correctly then everything which is painted by the application is subject to this scaling, and I haven't found any information about how to deactivate it just for some screen elements. Is the only way to somehow recognise the GDK_SCALE parameter and then try to paint everything with a scale factor of 0.5?

You understand correctly, as far as I know. As of Java 9, then it is possible to switch off the scaling — JOSM does so somehow.

I was about to say I had no idea how to do it, since I looked last week. But following [this method](https://stackoverflow.com/questions/43057457/jdk-9-high-dpi-disable-for-specific-panel/46630710#46630710) has given essentially the same result as JOSM: the 256px tiles shown sharp, although with the text at half the intended size. 

I added these lines to `MapCanvas#paint(Graphics inG)`:

```java
		final Graphics2D g = (Graphics2D) inG;
		final AffineTransform t = g.getTransform();
		final double scaling = t.getScaleX(); // Assuming square pixels :P
		t.setToScale(1, 1);
		g.setTransform(t);
```

I haven't done the "correct dimensions" bit, so the whole canvas isn't filled. But:

![screenshot_20180123_012743](https://user-images.githubusercontent.com/583320/35251816-67ddabc0-ffdd-11e7-83f2-880f09f1b55e.png)

:-D

> are there any public "official" tile servers producing 256x256 tiles with "extra-large-font-size" / "enhanced accessibility" ?

No, as above.

But I will make one later this week, and test using it.

**Comment by activityworkshop** [comment]
I haven't forgotten about this issue, and I now have access to some hardware where the problem is easily reproducible, so I can do some experimenting myself.
In my case, I'm not using a hardwired "2-times" scaling, requiring a 512x512 pixel image, but I've got a variable scaling factor which I can choose to be 100%, 125%, 150% or 175%.  I think this covers a good range rather than just assuming only either 100% or 200%.
I've taken your example and extended it a little, so this picture shows where I am now:

![image](https://user-images.githubusercontent.com/11003522/144289077-f16f5851-5959-4cc4-b195-27f5365da0f1.png)

So with 100% the menu and other GUI elements can be a bit small to read, and increasing the scaling factor makes them bigger but doesn't blur the map, just like you said.  The toolbar icons and zoom buttons get a bit blurred, of course.  The problem comes from the effect this has on converting from coordinates to pixels (for displaying things) and from pixels to coordinates (for drawing, panning and moving things).  So there's quite a bit of testing to do and perhaps some more lurking issues.  But if it seems reliable then this fix could go into 21.1.

What I'm still missing is the tiles (in 256x256 pixel format) using thick lines and large fonts - if these still aren't freely available then the map isn't going to be readable if you have tiny pixels.  Kind of surprising that these are still lacking!

**Comment by MattBlissett** [comment]
Excellent!

The layer switcher on https://www.openstreetmap.org/ has some high-resolution tiles (Cycle Map, Transport Map) but they are both 512x512px, as are the ones I made on https://tile.gbif.org/.  

I think you might be out of luck for 256x256.  With it being four times as many HTTP requests for a typical map, I doubt there's much demand for them — the web/Javascript libraries showing tiled maps generally want 256 pixel tiles for a standard display, and 512 pixel tiles for a double-density display, with some scaling in-between.

One option would be to make a small proxy, extracting a 256px tile from a 512px tile, which could be enough for testing the approach.

**Comment by activityworkshop** [comment]
I couldn't find the "high-resolution tiles (Cycle Map, Transport Map)" you mentioned on openstreetmap.org - for me they just look like normal 256x256 tiles as always.  Maybe you see something different using your browser?
The only tiles I could find with large renderings was at basemap.at which only covers Austria, so not useful for everybody.  But just as an experiment, I took two tiles from zoom 14, cut them into 4 tiles each and deployed them to zoom level 15, and this is the result:
![basemap_at](https://user-images.githubusercontent.com/11003522/144645434-bd94f89c-8802-4e11-b694-21af39b8e848.png)
That's the kind of rendering I was expecting, with none of that "some scaling in-between" that was the start of this whole discussion.  If there's a worldwide (freely available) tile source with this kind of rendering then we could be getting closer!

**Comment by MattBlissett** [comment]
The Cycle Map tiles shown on OpenStreetMap.org aren't necessarily free to use, information here: https://www.thunderforest.com/maps/ and an example tile: https://c.tile.thunderforest.com/cycle/10/547/319@2x.png?apikey=6170aad10dfd42a38d4d8c709a536f38 -- or maybe the "Hobby project" level would cover usage? Though I'd expect normal Javascript to use these when you look at the map: https://www.openstreetmap.org/#map=10/55.5997/12.1523&layers=C

OSMAnd has hi-resolution tiles, but without the cycling emphasis: https://tile.osmand.net/hd/10/547/319.png -- I can't find any details on free usage.

**Comment by activityworkshop** [comment]
Thanks to the OsmAnd people (and thanks to your tip MattBlissett!), I now have some 512-by-512 pixel tiles to play with, using that tile.osmand.net server you linked to.  Discussion is here: https://github.com/osmandapp/OsmAnd/discussions/13565.

So now it looks like this could go into GpsPrune version 21.2, with no new settings, no new dialogs, no new messages (so it doesn't have to wait for v22).

The following picture shows:
(1) 100% screen scaling by the OS, regular Mapnik tiles
(2) 125% screen scaling by the OS (note larger labels, but map is not zoomed and still sharp)
(3) 150% screen scaling by the OS, but now switched to the OsmAnd tile server and its 512-pixel tiles - so texts on the map tiels are rendered larger but still 100% sharp
(4) 175% screen scaling by the OS, and again the map is not zoomed and still sharp.

![fourzooms](https://user-images.githubusercontent.com/11003522/154855983-1becdb68-f2e9-474f-8bdd-9d27aeb44355.png)

would be grateful for feedback :)

**Comment by activityworkshop** [comment]
`finally {}`
:)
It's taken a while, but I believe this issue can be closed now, with GpsPrune 21.2.  For testing you'll probably want tiles with larger rendering, as discussed above, these can be either 256- or 512-pixel tiles.
All testing and all feedback is welcome, but if something is not working correctly then please raise a separate issue for that.  Thanks!
